hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/StreamBlockInput.java

```diff
@@ -0,0 +1,777 @@
+      if (readBlock.hasDataBuffers()) {
+        buffers.addAll(BufferUtils.getReadOnlyByteBuffers(
+            readBlock.getDataBuffers().getBuffersList()));
+      } else {
+        throw new IOException("Unexpected error while reading chunk data " +
+            "from container. No data returned.");
+      }
+    }
+    return response.getReadChunk(0)
+        .getChunkData().getOffset();
+  }
+
+  /**
+   * Check if the buffers have any data remaining between the current
+   * position and the limit.
+   */
+  private boolean buffersHaveData() {
+    boolean hasData = false;
+    if (buffersAllocated()) {
+      int buffersLen = buffers.size();
+      while (bufferIndex < buffersLen) {
+        ByteBuffer buffer = buffers.get(bufferIndex);
+        if (buffer != null && buffer.hasRemaining()) {
+          // current buffer has data
+          hasData = true;
+          break;
+        } else {
+          if (bufferIndex < buffersLen - 1) {
+            // move to next available buffer
+            ++bufferIndex;
+            Preconditions.checkState(bufferIndex < buffers.size());
+          } else {
+            // no more buffers remaining
+            break;
+          }
+        }
+      }
+    }
+
+    return hasData;
+  }
+
+  /**
+   * Check if there is more data in the chunk which has not yet been read
+   * into the buffers.
+   */
+  private boolean dataRemainingInBlock() {
+    long bufferPos;
+    if (blockPosition >= 0) {
+      bufferPos = blockPosition;
+    } else {
+      bufferPos = bufferOffsetWrtBlockDataData + buffersSize;
+    }
+
+    return bufferPos < length;
+  }
+
+  /**
+   * Check if current buffer had been read till the end.
+   */
+  private boolean bufferEOF() {
+    return allocated && !buffers.get(bufferIndex).hasRemaining();
+  }
+
+  /**
+   * Release the buffers upto the given index.
+   * @param releaseUptoBufferIndex bufferIndex (inclusive) upto which the
+   *                               buffers must be released
+   */
+  private void releaseBuffers(int releaseUptoBufferIndex) {
+    int buffersLen = buffers.size();
+    if (releaseUptoBufferIndex == buffersLen - 1) {
+      // Before releasing all the buffers, if block EOF is not reached, then
+      // blockPosition should be set to point to the last position of the
+      // buffers. This should be done so that getPos() can return the current
+      // block position
+      blockPosition = bufferOffsetWrtBlockDataData +
+          bufferoffsets.get(releaseUptoBufferIndex) +
+          buffers.get(releaseUptoBufferIndex).capacity();
+      // Release all the buffers
+      releaseBuffers();
+    } else {
+      buffers = buffers.subList(releaseUptoBufferIndex + 1, buffersLen);
+      bufferoffsets = bufferoffsets.subList(
+          releaseUptoBufferIndex + 1, buffersLen);
+    }
+  }
+
+  /**
+   * If EOF is reached, release the buffers.
+   */
+  private void releaseBuffers() {
+    buffers = null;
```

Comment:
Would `buffers.clear()` be better ?

Reply:
I think `buffers = null` is good. While reading new data, we use new buffer list from container. We don't reuse the buffer list, so I just let it clean up by GC.